### PR TITLE
Generalize Cluster Resources

### PR DESCRIPTION
**Pull request by andrewsykim**
Did some refactoring to further generalize any cloud provider implementations. Only renamed files + structs, there should be no changes in any of the cluster deletion logic (though it doesn't hurt to double check in the review).

Related issues: https://github.com/kubernetes/kops/issues/2150, https://github.com/kubernetes/kops/issues/2310

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2645)

### CODE REVIEW

**Comment by andrewsykim** [comment]
related: https://github.com/kubernetes/kops/issues/2646

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

File: cmd/kops/delete_cluster.go [hunk 1]
```diff
@@ -143,7 +142,7 @@ func RunDeleteCluster(f *util.Factory, out io.Writer, options *DeleteClusterOpti
 		}
 
 		// Todo lets make this smart enough to detect the cloud and switch on the ClusterResources interface
```
Comment (chrislovecnm) [inline cmd/kops/delete_cluster.go]: Can we remove the todo?

Comment (chrislovecnm) [inline cmd/kops/delete_cluster.go]: Pure nit pick, your call ... looks great!

**Comment by andrewsykim** [comment]
@chrislovecnm I didn't do too much testing as I only moved some files around and renamed a few things. I also thought cluster creation/deletion on the e2e tests would be enough. I can do some manual verification (or add more tests?) if you'd like :)

**Comment by andrewsykim** [comment]
I should have some bandwidth to do some more testing on this actually, should take a day or two :)

**Comment by chrislovecnm** [comment]
@andrewsykim please do some more testing 😄 

I will mark as WIP

**Comment by chrislovecnm** [comment]
Going to close and reopen so CI will run.

**Comment by andrewsykim** [comment]
Cluster creation:
```
AndrewSyKim bash~/go/src/k8s.io/kops $ kops version
Version 1.6.0 (git-e6cb451)
AndrewSyKim bash~/go/src/k8s.io/kops $ kops create cluster --cloud=gce --name=dev.asykim.com --zones=us-central1-a --project=andrew-dev-169300 --yes
I0530 20:51:12.979887   46704 create_cluster.go:816] Using SSH public key: /Users/AndrewSyKim/.ssh/id_rsa.pub
I0530 20:51:12.980013   46704 subnets.go:183] Assigned CIDR 172.20.32.0/19 to subnet us-central1-a
I0530 20:51:12.980948   46704 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420dbcfc0), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc420192000)}
I0530 20:51:12.980979   46704 clouddns.go:100] Successfully got DNS service: &{0xc420dbd020 https://www.googleapis.com/dns/v1/projects/  0xc420227290 0xc420227298 0xc4202272a0 0xc4202272a8}
W0530 20:51:13.262117   46704 apiserver.go:90] Enabling apiserver insecure port, for healthchecks (issue #43784)
W0530 20:51:13.262800   46704 apiserver.go:90] Enabling apiserver insecure port, for healthchecks (issue #43784)
I0530 20:51:16.735729   46704 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420d98840), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc4207206c0)}
I0530 20:51:16.735752   46704 clouddns.go:100] Successfully got DNS service: &{0xc420d988a0 https://www.googleapis.com/dns/v1/projects/  0xc4216e65d0 0xc4216e65d8 0xc4216e65e0 0xc4216e65e8}
W0530 20:51:16.941850   46704 apiserver.go:90] Enabling apiserver insecure port, for healthchecks (issue #43784)
W0530 20:51:16.942501   46704 apiserver.go:90] Enabling apiserver insecure port, for healthchecks (issue #43784)
I0530 20:51:18.048957   46704 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420e5f5f0), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc420b25440)}
I0530 20:51:18.048990   46704 clouddns.go:100] Successfully got DNS service: &{0xc420e5f680 https://www.googleapis.com/dns/v1/projects/  0xc4202262a0 0xc4202262b0 0xc4202262b8 0xc4202262c8}
W0530 20:51:18.359299   46704 external_access.go:35] TODO: Harmonize gcemodel ExternalAccessModelBuilder with awsmodel
W0530 20:51:18.359337   46704 firewall.go:34] TODO: Harmonize gcemodel with awsmodel for firewall - GCE model is way too open
W0530 20:51:18.359345   46704 firewall.go:61] Adding overlay network for X -> node rule - HACK
W0530 20:51:18.359348   46704 firewall.go:62] We should probably use subnets?
W0530 20:51:18.359366   46704 firewall.go:112] Adding overlay network for X -> master rule - HACK
I0530 20:51:19.674170   46704 executor.go:91] Tasks: 0 done / 40 total; 27 can run
I0530 20:51:20.312500   46704 vfs_castore.go:422] Issuing new certificate: "master"
I0530 20:51:20.318407   46704 vfs_castore.go:422] Issuing new certificate: "kube-proxy"
I0530 20:51:20.358832   46704 vfs_castore.go:422] Issuing new certificate: "kubelet"
I0530 20:51:20.466767   46704 vfs_castore.go:422] Issuing new certificate: "kops"
I0530 20:51:20.509995   46704 vfs_castore.go:422] Issuing new certificate: "kube-scheduler"
I0530 20:51:20.562285   46704 vfs_castore.go:422] Issuing new certificate: "kube-controller-manager"
I0530 20:51:20.660110   46704 vfs_castore.go:422] Issuing new certificate: "kubecfg"
I0530 20:51:24.737529   46704 executor.go:91] Tasks: 27 done / 40 total; 11 can run
I0530 20:51:24.875336   46704 instancetemplate.go:201] We should be using NVME for GCE
I0530 20:51:24.875875   46704 instancetemplate.go:201] We should be using NVME for GCE
I0530 20:51:25.049283   46704 instancetemplate.go:201] We should be using NVME for GCE
I0530 20:51:25.066461   46704 instancetemplate.go:201] We should be using NVME for GCE
I0530 20:51:29.182486   46704 executor.go:91] Tasks: 38 done / 40 total; 2 can run
I0530 20:51:33.356837   46704 executor.go:91] Tasks: 40 done / 40 total; 0 can run
I0530 20:51:33.356888   46704 dns.go:152] Pre-creating DNS records
I0530 20:51:33.357007   46704 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420bb1c20), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc420b51bc0)}
I0530 20:51:33.357043   46704 clouddns.go:100] Successfully got DNS service: &{0xc420bb1c80 https://www.googleapis.com/dns/v1/projects/  0xc4200329a0 0xc4200329a8 0xc4200329b0 0xc4200329b8}
I0530 20:51:34.252668   46704 update_cluster.go:229] Exporting kubecfg for cluster
Kops has set your kubectl context to dev.asykim.com

Cluster is starting.  It should be ready in a few minutes.

Suggestions:
 * validate cluster: kops validate cluster
 * list nodes: kubectl get nodes --show-labels
 * ssh to the master: ssh -i ~/.ssh/id_rsa admin@api.dev.asykim.com
The admin user is specific to Debian. If not using Debian please use the appropriate user based on your OS.
 * read about installing addons: https://github.com/kubernetes/kops/blob/master/docs/addons.md
```

Cluster Deletion:
```
AndrewSyKim bash~/go/src/k8s.io/kops $ kops version
Version 1.6.0 (git-e6cb451)
AndrewSyKim bash~/go/src/k8s.io/kops $ kops delete cluster dev.asykim.com --yes
I0530 20:59:28.409819   46710 gce.go:80] Scanning zones: [us-central1-a us-central1-b us-central1-c us-central1-f]
I0530 20:59:30.478370   46710 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420da1380), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc4206aa000)}
I0530 20:59:30.478411   46710 clouddns.go:100] Successfully got DNS service: &{0xc420da13e0 https://www.googleapis.com/dns/v1/projects/  0xc4202ac0d8 0xc4202ac0e0 0xc4202ac0e8 0xc4202ac0f0}
TYPE			NAME							ID
DNS Zone		asykim.com.						asykim.com.
Disk			a-etcd-events-dev-asykim-com				a-etcd-events-dev-asykim-com
Disk			a-etcd-main-dev-asykim-com				a-etcd-main-dev-asykim-com
Instance		master-us-central1-a-pdbm				us-central1-a/master-us-central1-a-pdbm
Instance		nodes-b7jn						us-central1-a/nodes-b7jn
Instance		nodes-s058						us-central1-a/nodes-s058
InstanceGroupManager	us-central1-a-master-us-central1-a-dev-asykim-com	us-central1-a/us-central1-a-master-us-central1-a-dev-asykim-com
InstanceGroupManager	us-central1-a-nodes-dev-asykim-com			us-central1-a/us-central1-a-nodes-dev-asykim-com
InstanceTemplate	master-us-central1-a-dev-asykim-com-1496191885		master-us-central1-a-dev-asykim-com-1496191885
InstanceTemplate	nodes-dev-asykim-com-1496191885				nodes-dev-asykim-com-1496191885
Route			dev-asykim-com-8a5ba54d-459b-11e7-9fee-42010a800004	dev-asykim-com-8a5ba54d-459b-11e7-9fee-42010a800004
Route			dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004	dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004
Route			dev-asykim-com-daaedb7d-459b-11e7-9fee-42010a800004	dev-asykim-com-daaedb7d-459b-11e7-9fee-42010a800004

DNS Zone:asykim.com.	ok
Route:dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004	ok
Route:dev-asykim-com-daaedb7d-459b-11e7-9fee-42010a800004	ok
Route:dev-asykim-com-8a5ba54d-459b-11e7-9fee-42010a800004	ok
Instance:us-central1-a/nodes-s058	ok
Instance:us-central1-a/master-us-central1-a-pdbm	ok
Instance:us-central1-a/nodes-b7jn	ok
I0530 21:00:31.841001   46710 op.go:121] waitForOperation: long operation (1m0.163382511s): {"endTime":"2017-05-30T18:00:31.092-07:00","id":"6838096234000133020","insertTime":"2017-05-30T17:59:31.487-07:00","kind":"compute#operation","name":"operation-1496192371112-550c76ecbb840-23edd075-4f660c77","operationType":"compute.instanceGroupManagers.delete","progress":100,"selfLink":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a/operations/operation-1496192371112-550c76ecbb840-23edd075-4f660c77","startTime":"2017-05-30T17:59:31.780-07:00","status":"DONE","targetId":"4814096713986059646","targetLink":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a/instanceGroupManagers/us-central1-a-nodes-dev-asykim-com","user":"kim.andrewsy@gmail.com","zone":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a"}
InstanceGroupManager:us-central1-a/us-central1-a-nodes-dev-asykim-com	ok
I0530 21:00:34.952079   46710 op.go:121] waitForOperation: long operation (1m3.148823096s): {"endTime":"2017-05-30T18:00:33.839-07:00","id":"3734491400633559964","insertTime":"2017-05-30T17:59:31.565-07:00","kind":"compute#operation","name":"operation-1496192371112-550c76ecbb841-fff423a9-2e59e80d","operationType":"compute.instanceGroupManagers.delete","progress":100,"selfLink":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a/operations/operation-1496192371112-550c76ecbb841-fff423a9-2e59e80d","startTime":"2017-05-30T17:59:31.931-07:00","status":"DONE","targetId":"6298128173838347646","targetLink":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a/instanceGroupManagers/us-central1-a-master-us-central1-a-dev-asykim-com","user":"kim.andrewsy@gmail.com","zone":"https://www.googleapis.com/compute/beta/projects/andrew-dev-169300/zones/us-central1-a"}
InstanceGroupManager:us-central1-a/us-central1-a-master-us-central1-a-dev-asykim-com	ok
InstanceTemplate:master-us-central1-a-dev-asykim-com-1496191885	ok
InstanceTemplate:nodes-dev-asykim-com-1496191885	ok
Not all resources deleted; waiting before reattempting deletion
	Disk:a-etcd-main-dev-asykim-com
	Disk:a-etcd-events-dev-asykim-com
Not all resources deleted; waiting before reattempting deletion
	Disk:a-etcd-events-dev-asykim-com
	Disk:a-etcd-main-dev-asykim-com
Not all resources deleted; waiting before reattempting deletion
	Disk:a-etcd-main-dev-asykim-com
	Disk:a-etcd-events-dev-asykim-com
Not all resources deleted; waiting before reattempting deletion
	Disk:a-etcd-main-dev-asykim-com
	Disk:a-etcd-events-dev-asykim-com
Not all resources deleted; waiting before reattempting deletion
	Disk:a-etcd-events-dev-asykim-com
	Disk:a-etcd-main-dev-asykim-com
^C
AndrewSyKim bash~/go/src/k8s.io/kops $ kops delete cluster dev.asykim.com --yes
I0530 21:01:24.777503   46711 gce.go:80] Scanning zones: [us-central1-a us-central1-b us-central1-c us-central1-f]
I0530 21:01:26.563602   46711 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:(*oauth2.tokenRefresher)(0xc420d87b60), mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(0xc4206e4cc0)}
I0530 21:01:26.563639   46711 clouddns.go:100] Successfully got DNS service: &{0xc420d87bc0 https://www.googleapis.com/dns/v1/projects/  0xc42054c598 0xc42054c5a0 0xc42054c5a8 0xc42054c5b0}
TYPE		NAME							ID
DNS Zone	asykim.com.						asykim.com.
Disk		a-etcd-events-dev-asykim-com				a-etcd-events-dev-asykim-com
Disk		a-etcd-main-dev-asykim-com				a-etcd-main-dev-asykim-com
Route		dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004	dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004

DNS Zone:asykim.com.	ok
Disk:a-etcd-main-dev-asykim-com	ok
Disk:a-etcd-events-dev-asykim-com	ok
Route:dev-asykim-com-d4b425a7-459b-11e7-9fee-42010a800004	ok
Deleted kubectl config for dev.asykim.com

Cluster deleted
```

**Comment by andrewsykim** [comment]
There seems to be a bug where disks on GCE aren't deleted, retrying delete cluster seems to do the job. I can reproduce this on master so I don't think its related (will open another issue for this later).

**Comment by chrislovecnm** [comment]
Waiting on CI to pass before LGTM

**Pull request merged**